File: tools/workout_history.py
#!/usr/bin/env python3
"""Maintain a private local Workout Inspector history from FIT/ZIP exports."""

from __future__ import annotations

import argparse
import hashlib
import json
from datetime import datetime, timezone
from pathlib import Path
from statistics import median

try:
    from validate_workout import load_report
except ModuleNotFoundError:
    from tools.validate_workout import load_report


METRICS = {
    "duration": "s",
    "work_seconds": "s",
    "sets": "sets",
    "motion_exposure": "mg-s",
    "active_seconds": "s",
    "weight_volume": "kg-swings",
    "motion_peak": "mg",
    "smoothness": "score",
}


def metric(laps: list[dict], name: str) -> float | None:
    values = [float(lap[name]) for lap in laps if isinstance(lap.get(name), (int, float)) and lap[name] >= 0]
    if not values:
        return None
    if name == "motion_peak":
        return max(values)
    if name == "smoothness":
        return sum(values) / len(values)
    return sum(values)


def summarize(report: dict, identifier: str, imported_at: str | None = None) -> dict:
    work = [lap for lap in report["laps"] if lap.get("phase") == "work" and lap.get("set", 0) > 0]
    return {
        "id": identifier,
        "imported_at": imported_at or datetime.now(timezone.utc).isoformat(),
        "date": report["summary"].get("date"),
        "movement": report["summary"].get("movement"),
        "side": report["summary"].get("side"),
        "equipment": report["summary"].get("equipment"),
        "sets": len(work),
        "duration": report["summary"].get("elapsed"),
        "metrics": {
            "duration": report["summary"].get("elapsed"),
            "work_seconds": report["summary"].get("work_seconds"),
            "sets": len(work),
            **{name: metric(work, name) for name in METRICS
               if name not in ("duration", "work_seconds", "sets")},
        },
    }


def comparisons(entries: list[dict], minimum: int = 3, window: int = 8) -> list[dict]:
    if not entries:
        return []
    current = entries[-1]
    previous = [item for item in entries[max(0, len(entries) - 1 - window):-1]
                if (item.get("metrics", {}).get("sets") or 0) > 0
                and (not current.get("movement") or current.get("movement") == "Unknown"
                     or item.get("movement") == current.get("movement"))
                and (not current.get("equipment") or item.get("equipment") == current.get("equipment"))]
    results = []
    for name, unit in METRICS.items():
        value = current["metrics"].get(name)
        baseline = [item["metrics"].get(name) for item in previous]
        baseline = [item for item in baseline if isinstance(item, (int, float)) and item > 0]
        if value is None or len(baseline) < minimum:
            results.append({"metric": name, "status": "unavailable", "samples": len(baseline)})
            continue
        typical = median(baseline)
        change = (value - typical) / typical * 100 if typical else None
        direction = "higher" if change is not None and change >= 25 else "lower" if change is not None and change <= -25 else "typical"
        results.append({
            "metric": name, "status": "available", "value": round(value, 1), "unit": unit,
            "baseline": round(typical, 1), "change_percent": round(change, 1) if change is not None else None,
            "direction": direction, "review": change is not None and abs(change) >= 50,
            "samples": len(baseline),
        })
    return results


def load(path: Path) -> dict:
    if not path.exists():
        return {"version": 1, "privacy": "local-summary-only", "workouts": []}
    return json.loads(path.read_text())


def save(path: Path, history: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(history, indent=2) + "\n")


def source_id(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()[:12]


def import_source(history: dict, source: Path) -> dict:
    identifier = source_id(source)
    if any(item["id"] == identifier for item in history["workouts"]):
        return history
    history["workouts"].append(summarize(load_report(source), identifier))
    history["workouts"].sort(key=lambda item: datetime.strptime(item["date"], "%d %b %Y")
                             if item.get("date") else datetime.min)
    return history


def parse_args(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--database", type=Path, default=Path("workout-history.json"))
    sub = parser.add_subparsers(dest="command", required=True)
    add = sub.add_parser("import"); add.add_argument("sources", nargs="+", type=Path)
    sub.add_parser("show")
    export = sub.add_parser("export"); export.add_argument("output", type=Path)
    delete = sub.add_parser("delete"); delete.add_argument("id")
    sub.add_parser("clear")
    return parser.parse_args(argv)


def main(argv=None) -> int:
    args = parse_args(argv)
    history = load(args.database)
    if args.command == "import":
        for source in args.sources:
            import_source(history, source)
        save(args.database, history)
    elif args.command == "delete":
        history["workouts"] = [item for item in history["workouts"] if item["id"] != args.id]
        save(args.database, history)
    elif args.command == "clear":
        history["workouts"] = []
        save(args.database, history)
    elif args.command == "export":
        save(args.output, history)
    output = {**history, "latest_comparison": comparisons(history["workouts"])}
    print(json.dumps(output, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/workout-history.js
#!/usr/bin/env node
// Maintain a private local Workout Inspector history from FIT/ZIP exports.

import { createHash } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { pythonRound } from "./fit-io.js";
import { loadReport } from "./validate-workout.js";

const METRICS = {
    duration: "s",
    work_seconds: "s",
    sets: "sets",
    motion_exposure: "mg-s",
    active_seconds: "s",
    weight_volume: "kg-swings",
    motion_peak: "mg",
    smoothness: "score",
};

const MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

function median(values) {
    const sorted = [...values].sort((a, b) => a - b);
    const middle = Math.floor(sorted.length / 2);
    return sorted.length % 2 ? sorted[middle] : (sorted[middle - 1] + sorted[middle]) / 2;
}

export function metric(laps, name) {
    const values = laps
        .filter((lap) => typeof lap[name] === "number" && lap[name] >= 0)
        .map((lap) => Number(lap[name]));
    if (values.length === 0) {
        return null;
    }
    if (name === "motion_peak") {
        return Math.max(...values);
    }
    if (name === "smoothness") {
        return values.reduce((total, value) => total + value, 0) / values.length;
    }
    return values.reduce((total, value) => total + value, 0);
}

export function summarize(report, identifier, importedAt = null) {
    const work = report.laps.filter((lap) => lap.phase === "work" && (lap.set ?? 0) > 0);
    const metrics = { duration: report.summary.elapsed ?? null, work_seconds: report.summary.work_seconds ?? null, sets: work.length };
    for (const name of Object.keys(METRICS)) {
        if (!(name in metrics)) {
            metrics[name] = metric(work, name);
        }
    }
    return {
        id: identifier,
        imported_at: importedAt ?? new Date().toISOString(),
        date: report.summary.date ?? null,
        movement: report.summary.movement ?? null,
        side: report.summary.side ?? null,
        equipment: report.summary.equipment ?? null,
        sets: work.length,
        duration: report.summary.elapsed ?? null,
        metrics,
    };
}

export function comparisons(entries, minimum = 3, window = 8) {
    if (entries.length === 0) {
        return [];
    }
    const current = entries[entries.length - 1];
    const previous = entries.slice(Math.max(0, entries.length - 1 - window), -1).filter(
        (item) => (item.metrics?.sets || 0) > 0
            && (!current.movement || current.movement === "Unknown" || item.movement === current.movement)
            && (!current.equipment || item.equipment === current.equipment),
    );
    const results = [];
    for (const [name, unit] of Object.entries(METRICS)) {
        const value = current.metrics?.[name];
        const baseline = previous
            .map((item) => item.metrics?.[name])
            .filter((item) => typeof item === "number" && item > 0);
        if (value == null || baseline.length < minimum) {
            results.push({ metric: name, status: "unavailable", samples: baseline.length });
            continue;
        }
        const typical = median(baseline);
        const change = typical ? ((value - typical) / typical) * 100 : null;
        const direction = change !== null && change >= 25 ? "higher"
            : change !== null && change <= -25 ? "lower" : "typical";
        results.push({
            metric: name, status: "available", value: pythonRound(value, 1), unit,
            baseline: pythonRound(typical, 1),
            change_percent: change !== null ? pythonRound(change, 1) : null,
            direction, review: change !== null && Math.abs(change) >= 50,
            samples: baseline.length,
        });
    }
    return results;
}

export function load(path) {
    if (!existsSync(path)) {
        return { version: 1, privacy: "local-summary-only", workouts: [] };
    }
    return JSON.parse(readFileSync(path, "utf8"));
}

export function save(path, history) {
    mkdirSync(dirname(path) || ".", { recursive: true });
    writeFileSync(path, `${JSON.stringify(history, null, 2)}\n`, "utf8");
}

export function sourceId(path) {
    return createHash("sha256").update(readFileSync(path)).digest("hex").slice(0, 12);
}

function dateKey(item) {
    const match = item.date ? /^(\d{2}) ([A-Z][a-z]{2}) (\d{4})$/.exec(item.date) : null;
    if (!match || !MONTHS.includes(match[2])) {
        return -Infinity;
    }
    return Date.UTC(Number(match[3]), MONTHS.indexOf(match[2]), Number(match[1]));
}

export function importSource(history, source, loader = loadReport) {
    const identifier = sourceId(source);
    if (history.workouts.some((item) => item.id === identifier)) {
        return history;
    }
    history.workouts.push(summarize(loader(source), identifier));
    history.workouts.sort((a, b) => dateKey(a) - dateKey(b));
    return history;
}

function parseArgs(argv) {
    const args = { database: "workout-history.json", command: null, sources: [], output: null, id: null };
    let index = 0;
    while (index < argv.length) {
        const argument = argv[index];
        if (argument === "--database") {
            args.database = argv[++index];
        } else if (args.command === null) {
            args.command = argument;
        } else {
            args.sources.push(argument);
        }
        index++;
    }
    if (!["import", "show", "export", "delete", "clear"].includes(args.command)) {
        throw new Error("usage: workout-history.js [--database PATH] {import|show|export|delete|clear} ...");
    }
    if (args.command === "import" && args.sources.length === 0) {
        throw new Error("import requires at least one FIT/ZIP source");
    }
    if (args.command === "export") {
        args.output = args.sources[0] ?? null;
        if (!args.output) {
            throw new Error("export requires an output path");
        }
    }
    if (args.command === "delete") {
        args.id = args.sources[0] ?? null;
        if (!args.id) {
            throw new Error("delete requires a workout id");
        }
    }
    return args;
}

export function main(argv = process.argv.slice(2), loader = loadReport) {
    const args = parseArgs(argv);
    const history = load(args.database);
    if (args.command === "import") {
        for (const source of args.sources) {
            importSource(history, source, loader);
        }
        save(args.database, history);
    } else if (args.command === "delete") {
        history.workouts = history.workouts.filter((item) => item.id !== args.id);
        save(args.database, history);
    } else if (args.command === "clear") {
        history.workouts = [];
        save(args.database, history);
    } else if (args.command === "export") {
        save(args.output, history);
    }
    console.log(JSON.stringify({ ...history, latest_comparison: comparisons(history.workouts) }, null, 2));
    return 0;
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
    process.exitCode = main();
}
